src/ray/raylet/scheduling/cluster_task_manager.cc

```diff
@@ -332,15 +323,91 @@ void ClusterTaskManager::TaskFinished(std::shared_ptr<WorkerInterface> worker,
     ReleaseWorkerResources(worker);
   }
 }
 
+bool ClusterTaskManager::PinTaskArgsIfMemoryAvailable(const TaskSpecification &spec,
+                                                      bool *args_missing) {
+  std::vector<std::unique_ptr<RayObject>> args;
+  const auto &deps = spec.GetDependencyIds();
+  if (!deps.empty()) {
+    // This gets refs to the arguments stored in plasma. The refs should be
+    // deleted once we no longer need to pin the arguments.
+    if (!pin_task_arguments_(deps, &args)) {
+      *args_missing = true;
+      return false;
+    }
+    for (size_t i = 0; i < deps.size(); i++) {
+      if (args[i] == nullptr) {
+        // This can happen if the task's arguments were all local at some
+        // point, but then at least one was evicted before the task could
+        // be dispatched to a worker.
+        RAY_LOG(INFO)
+            << "Task " << spec.TaskId() << " argument " << deps[i]
+            << " was evicted before the task could be dispatched. This can happen "
+               "when there are many objects needed on this node. The task will be "
+               "scheduled once all of its dependencies are local.";
+        *args_missing = true;
+        return false;
+      }
+    }
+  }
+
+  *args_missing = false;
+  size_t task_arg_bytes = 0;
+  for (auto &arg : args) {
+    task_arg_bytes += arg->GetSize();
+  }
+  PinTaskArgs(spec, std::move(args));
+  RAY_LOG(DEBUG) << "Size of pinned task args is now " << pinned_task_arguments_bytes_;
+  if (task_arg_bytes > max_pinned_task_arguments_bytes_) {
+    RAY_LOG(WARNING)
+        << "Dispatched task " << spec.TaskId() << " has arguments of size "
+        << task_arg_bytes
+        << ", but the max memory allowed for arguments of executing tasks is only "
+        << max_pinned_task_arguments_bytes_;
+  } else if (pinned_task_arguments_bytes_ > max_pinned_task_arguments_bytes_) {
+    ReleaseTaskArgs(spec.TaskId());
+    return false;
+  }
+
+  return true;
+}
+
+void ClusterTaskManager::PinTaskArgs(const TaskSpecification &spec,
+                                     std::vector<std::unique_ptr<RayObject>> args) {
+  const auto &deps = spec.GetDependencyIds();
+  for (size_t i = 0; i < deps.size(); i++) {
+    auto inserted =
+        pinned_task_arguments_.emplace(deps[i], std::make_pair(std::move(args[i]), 0));
+    auto it = inserted.first;
+    if (inserted.second) {
+      pinned_task_arguments_bytes_ += it->second.first->GetSize();
+    }
+    it->second.second++;
+  }
+  RAY_CHECK(executing_task_args_.emplace(spec.TaskId(), deps).second) << spec.TaskId();
+}
+
+void ClusterTaskManager::ReleaseTaskArgs(const TaskID &task_id) {
+  auto it = executing_task_args_.find(task_id);
+  RAY_CHECK(it != executing_task_args_.end());
+  for (auto &arg : it->second) {
+    auto arg_it = pinned_task_arguments_.find(arg);
+    RAY_CHECK(arg_it != pinned_task_arguments_.end());
+    RAY_CHECK(arg_it->second.second > 0);
+    arg_it->second.second--;
```

Comment:
Can you use something like 

```
auto &obj_ref_pair = arg_it->second; 
```

I think `arg_it->second.second` is a bit hard to read.